### DISCUSSION TITLE
MSBuild optional WithProperty

### DISCUSSION THREAD

**Discussion by nick5454** [cake-build/cake#3383] [Extension Q&A] [answered]
I have a MSBuild statement, but for `WithProperty` I only want to add 2 flags if `isAndroid` variable is `true`.

Something like `.WithProperty(test => IsAndroid == true, "AndroidBuildApplicationPackage", true)`

Is this possible, actual code is below:

```cakescript
MSBuild(BuildParameters.SolutionFile.ToString(), configurator => configurator
    .SetConfiguration(BuildParameters.Configuration)
    .WithProperty("Platform", BuildParameters.Platform)
    .SetVerbosity(Verbosity.Normal)
    .WithProperty("BuildIpa", "true")
    //.WithProperty("AndroidBuildApplicationPackage", "true")
    //.WithProperty("Project", "appname.Android")
    .WithProperty("CodesignKey",  BuildParameters.CertTypes[BranchProject].IOSTeam)
    .WithProperty("CodesignProvision", provisionType) 
    .WithProperty("MtouchNoSymbolStrip", "false")
    .WithProperty("ArchiveOnBuild", "true")
    .WithLogger(
        Context.Tools.Resolve("MSBuild.ExtensionPack.Loggers.dll").FullPath,
        "XmlFileLogger",
        $"logfile=\"{BuildParameters.LogPath}\";verbosity=Detailed;encoding=UTF-8"
    )
);
```

**Comment by augustoproiete** [comment, marked as answer]
There's nothing built-in at the moment that would give you a predicate to decide to add (or not) a property.

Is there any value of `AndroidBuildApplicationPackage` that would have the same effect as not having the property at all? e.g. `false`? If so, you could do something easy such as:

```cakescript
.WithProperty("AndroidBuildApplicationPackage", isAndroid ? "false" : "true")
```

You can see a [similar example here](https://github.com/augustoproiete/Cake.Args/blob/v1.0.1/build.cake#L37).

---

Otherwise, you'd need to create an instance of [`MSBuildSettings`](https://cakebuild.net/api/Cake.Common.Tools.MSBuild/MSBuildSettings/), call (or not call) `WithProperty` based on the condition/variable and then use the [`MSBuild` overload that takes an instance of `MSBuildSettings`](https://cakebuild.net/api/Cake.Common.Tools.MSBuild/MSBuildAliases/C240F0FB).

e.g.

```cakescript
var settings = new MSBuildSettings();
settings.WithProperty("Platform", BuildParameters.Platform);
// etc.

if (isAndroid)
{
    settings.WithProperty("AndroidBuildApplicationPackage", "true");
}

MSBuild(BuildParameters.SolutionFile.ToString(), settings);
```

**Reply by nick5454** [reply]
Silly me, I wasn't sure if by the presence of the flag that would change anything
